models/user.go

```diff
@@ -1 +1,337 @@
+	USER_CONTACT_NUMBER_MAX_LENGTH = 20
+)
+
+// User contains the details about the user.
+type User struct {
+	ID                      string            `json:"id"`
+	CreatedAt               int64             `json:"created_at,omitempty"`
+	UpdatedAt               int64             `json:"updated_at,omitempty"`
+	DeletedAt               int64             `json:"deleted_at"`
+	Username                string            `json:"username"`
+	Password                string            `json:"password,omitempty"`
+	AuthData                *string           `json:"auth_data,omitempty"`
+	AuthService             string            `json:"auth_service"`
+	Email                   string            `json:"email"`
+	EmailVerified           bool              `json:"email_verified,omitempty"`
+	FirstName               string            `json:"first_name"`
+	LastName                string            `json:"last_name"`
+	ContactNumber           string            `json:"contact_number"`
+	FacebookUserID          string            `json:"facebook_user_id"`
+	GoogleUserID            string            `json:"google_user_id"`
+	Locale                  string            `json:"locale"`
+	Timezone                map[string]string `json:"timezone"`
+	TermsOfServiceId        string            `db:"-" json:"terms_of_service_id,omitempty"`
+	TermsOfServiceCreatedAt int64             `db:"-" json:"terms_of_service_created_at,omitempty"`
+}
+
+type UserPatch struct {
+	Username  *string           `json:"username"`
+	Password  *string           `json:"password,omitempty"`
+	FirstName *string           `json:"first_name"`
+	LastName  *string           `json:"last_name"`
+	Email     *string           `json:"email"`
+	Locale    *string           `json:"locale"`
+	Timezone  map[string]string `json:"timezone"`
+	RemoteId  *string           `json:"remote_id"`
+}
+
+type UserAuth struct {
+	AuthData    *string `json:"auth_data,omitempty"`
+	AuthService string  `json:"auth_service,omitempty"`
+}
+
+var validUsernameChars = regexp.MustCompile(`^[a-z0-9\.\-_]+$`)
+
+// IsValid validates the user and returns an error if it isn't configured correctly.
+func (u *User) IsValid() *AppError {
+
+	if !IsValidId(u.ID) {
+		return InvalidUserError("id", "")
+	}
+
+	if u.CreatedAt == 0 {
+		return InvalidUserError("create_at", u.ID)
+	}
+
+	if u.UpdatedAt == 0 {
+		return InvalidUserError("update_at", u.ID)
+	}
+
+	if len(u.Email) > USER_EMAIL_MAX_LENGTH || u.Email == "" || !IsValidEmail(u.Email) {
+		return InvalidUserError("email", u.ID)
+	}
+
+	if utf8.RuneCountInString(u.FirstName) > USER_FIRST_NAME_MAX_RUNES {
+		return InvalidUserError("first_name", u.ID)
+	}
+
+	if utf8.RuneCountInString(u.LastName) > USER_LAST_NAME_MAX_RUNES {
+		return InvalidUserError("last_name", u.ID)
+	}
+
+	if u.AuthData != nil && *u.AuthData != "" && u.AuthService == "" {
+		return InvalidUserError("auth_data_type", u.ID)
+	}
+
+	if u.Password != "" && u.AuthData != nil && *u.AuthData != "" {
```

Comment:
In case we don't plan to store the AuthData then I believe it should be based on whether the google user id / facebook user id exists.